runtime/indent/xml.vim: fix indentation when tags start with an undrscore

ao2 opened the pull request:
When indenting XML containing elements whose name start with an underscore vim fails to get the indentation levels.

The issue can be reproduced with any inkscape extension .inx file, e.g.:
https://raw.githubusercontent.com/ChristianBecker/inkscape-android-export/master/android_export.inx
1. Open the file: `vim -u NONE -c "filetype plugin indent on" -c "syn on"  -c "set ft=xml" android_export.inx`
2. Try indenting it using `gg=G`.

The proposed patch fixes this by allowing underscores at the beginning of the tags when looking for open tags.

This can be done easily by using the "`head of word character`" atom (`\h`) instead of the "`alphabetic character`" atom (`\a`) in `b:xml_indent_open`.

See "`help \h`".

chrisbra commented:
have you contacted the runtime file maintainer?

ao2 commented:
I have, I sent him the patch but I didn't know if he was active so I posted the changes here too.
Now I will also point him to this pull request; and wait some days to see if he replies.

Thanks,
   Antonio

rakus commented:
I sent a mail to the maintainer about this bug in November and December last year. Then I forgot about it (fixed it locally). So I guess he is not active anymore.

While looking at it I asked my self why not fix it this way:
    let b:xml_indent_open = '.{-}<[^/]'

A lot of characters are allowed as first char in a XML-tag see: 
http://www.w3.org/TR/REC-xml/#NT-NameStartChar

Anyway I think the fix is just fine.

brammool commented:
> I sent a mail to the maintainer about this bug in November and
> December last year. Then I forgot about it (fixed it locally). So I
> guess he is not active anymore.
> 
> While looking at it I asked my self why not fix it this way:
>     let b:xml_indent_open = '.{-}<[^/]'
> 
> A lot of characters are allowed as first char in a XML-tag see: 
> http://www.w3.org/TR/REC-xml/#NT-NameStartChar
> 
> Anyway I think the fix is just fine.

I have not heard from Johannes Zellner for a long time.
Does someone want to take over maintaining the xml indent script?

## 

Why doesn't Tarzan have a beard?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

ao2 commented:
@rakus I'd stay with the strict rule and expand it if necessary.

If we wanted to support the official spec closely we could have something like:

```
let b:xml_indent_open = '.\{-}<[:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\U00010000-\U000EFFFF]'
```

However I am not sure if it's really worth it, even the vim xml syntax file doesn't go that far.

I would add the `:` tho.

rakus commented:
Antonio,

I also think, that it is not worth it.

Ralf

ao2 commented:
@rakus OK, I will just add the `:` then in another commit (I am pushing it tomorrow).

About taking over in maintaining the xml vim files, if nobody else steps up I guess I could do it.

Thanks,
   Antonio

ao2 commented:
OK, added the colon too and set myself as the maintainer of `runtime/indent/xml.vim`.

ao2 commented:
ping

k-takata commented:
I think you should move the name of previous maintainer to "Previous Maintainer:".
E.g. https://github.com/vim/vim/blob/master/runtime/indent/sh.vim#L4

ao2 commented:
@k-takata that's a good idea, I can do that, thanks.

I could amend the commit in which I take over to have a cleaner history but I have a doubt.

What is the preferred workflow on github when changing history? Can I `git push -f` in the same branch? Or should I create a new branch and send a new pull request?

Amending/rebasing **public** stuff is generally considered a bad practice in git.

Thanks,
   Antonio

vim-ml commented:
Am 2016-08-25 12:37, schrieb Antonio Ospite:

> @k-takata [1] that's a good idea, I can do that, thanks.
> 
> I could amend the commit in which I take over to have a cleaner
> history but I have a doubt.
> 
> What is the preferred workflow on github when changing history? Can I
> git push -f in the same branch? Or should I create a new branch and
> send a new pull request?
> 
> Amending/rebasing PUBLIC stuff is generally considered a bad practice
> in git.

Yeah, but for pull requests it is usually okay. However,
the usual workflow for runtime updates in Vim varies greatly
(see 
https://github.com/vim/vim/blob/master/CONTRIBUTING.md#syntax-indent-and-other-runtime-files)

Bram prefers getting updated runtime files sent by mail and
he commits it then to the repository.

Best,
Christian

ao2 commented:
OK, I force-pushed the change suggested by @k-takata and added a new commit where I reformat the header to be more readable.

@brammool if you prefer a single patch sent over to the ml just say so, on my part I'd prefer having the single commits in the git history, but it's not a big deal.

Thanks,
   Antonio

brammool commented:
I prefer you send me the whole updated file.

ao2 committed:
runtime/indent/xml.vim: fix xml indentation and other updates

Changes since the previous version:

  - fix indentation when tags start with an underscore or a colon

    The XML spec says that element names can start with a colon too
    (among other characters):
    https://www.w3.org/TR/REC-xml/#NT-NameStartChar

    However when indenting XML containing elements whose name start with
    an underscore or a colon vim fails to get the indentation levels.

    So, allow underscores and colons in b:xml_indent_open, but in order
    to do that, the \a atom needs to be expanded to its equivalent
    characters class before adding ':' and '_' to it because character
    class cannot contain atoms (excepts some special ones), see the NOTE
    at the end of "help \[", (i.e.  [:_\a] would not have been valid).

  - make Antonio Ospite the new maintainer

    Johannes Zellner seems to have been unresponsive from quite some
    time, make Antonio Ospite the new maintainer of
    runtime/indent/xml.vim.

  - reformat the header to make it more readable

ao2 commented:
Ideally, with git, we should not be afraid of micro-commits, however I squashed all the commits into a single one, and also rebased on the current master.

If you don't plan to merge the pull request as is (so that I show up as the author in the git history too, <rant>that's how git is ideally meant to be used</rant>) I will send the whole file as you requested.

I just wanted to test the waters once again 😄.

Thanks,
   Antonio

EDIT: the original version sounded harsher than I intended, I just wanted to express my opinion on the vim workflow.

ao2 commented:
@chrisbra I see you are the xml runtime maintainer now, I will send the changes over to https://github.com/chrisbra/vim-xml-ftplugin

ao2 commented:
Closing this pull request, as the issue is now tracked in https://github.com/chrisbra/vim-xml-ftplugin/pull/2
